Modules/POSIX/Sources/POSIX/Regex.swift

```diff
@@ -0,0 +1,319 @@
+        try! self.init(pattern: value, options: .extended)
+    }
+}
+*/
+
+/// MARK: - Operators
+
+
+/// Check if a string matches a given Regex.
+///
+/// - parameter string: String to match.
+/// - parameter regex:  Regex to use for matching.
+///
+/// - returns: `true` if the given string matches the Regex, else `false`.
+public func ~ (string: String, regex: Regex) -> Bool {
+    return regex.matches(string)
+}
+
+/// Check if a string matches a given regex pattern.
+///
+/// - parameter string:  String to match.
+/// - parameter pattern: String holding the regex pattern.
+///
+/// - throws: `RegexError` if the given regex pattern as not a valid Regex.
+///
+/// - returns: `true` if the given string matches the regex pattern, else `false`.
+public func ~ (string: String, pattern: String) throws -> Bool {
+    let regex = try Regex(pattern: pattern)
+    return string ~ regex
+}
+
+/// Check if a string matches a given regex pattern.
+///
+/// - note: returns `nil` instead of throwing if the given regex pattern as not a valid Regex.
+///
+/// - parameter string:  String to match.
+/// - parameter pattern: String holding the regex pattern.
+///
+/// - returns: `true` if the given string matches the regex pattern, else `false`.
+public func ~? (string: String, pattern: String) -> Bool? {
+    return try? (string ~ pattern)
+}
+
+
+/// Matching groups in a string, given a regex pattern.
+///
+/// This function throws if a valid regex cannot be built out of the given pattern.
+///
+/// - parameter string:  String in which to search for groups.
+/// - parameter pattern: String pattern to use as a regex.
+///
+/// - throws: `RegexError` indicating why the given pattern is not a valid Regex.
+///
+/// - returns: Matching groups as an array of strings
+public func ~* (string: String, pattern: String) throws -> [String] {
+    let regex = try Regex(pattern: pattern)
+    return string ~* regex
+}
+
+/// Matching groups in a string, given a Regex.
+///
+/// - parameter string: String in which to search for groups.
+/// - parameter regex:  String pattern to use as a regex.
+///
+/// - returns: Matching groups as an array of strings
+public func ~* (string: String, regex: Regex) -> [String] {
+    return regex.groups(in: string)
+}
+
+/// Matching groups in a string, given a regex pattern.
+///
+/// - note: returns `nil` instead of throwing if the given regex pattern as not a valid Regex.
+///
+/// - parameter string:  String in which to search for groups.
+/// - parameter pattern: String pattern to use as a regex.
+///
+/// - returns: Matching groups as an array of strings, or `nil` if the given pattern is not a valid Regex.
+public func ~*? (string: String, pattern: String) -> [String]? {
+    guard let regex = try? Regex(pattern: pattern) else {
+        return nil
+    }
+    return string ~* regex
+}
+
+// FIXME: let's talk about this
+//func ~* <Result> (left: String, right: String) throws -> ((String) -> Result) -> [Result] {
```

Comment:
Yep, it's just syntactic sugar for map or closure:

For the operators, I intend to adopt the ruby operator syntax `=~`

```
try ("hello world" =~ "(hello) (world)") { match in 
    return match.contains("wo") ? match.uppercased() : match  // Result = String
}
```

or 
```
try ("hello world" =~ "(hello) world") { match in
    print(match)  // Result inferred to be Void
}
```

but I now think that adding `.map` does not hurt :)

BTW, I tried to have this working without the parens but I could not find a good precedence group : the expressions above, when removing the parens, are understood as "the closure is applied to the 2nd string", which fails miserably :)